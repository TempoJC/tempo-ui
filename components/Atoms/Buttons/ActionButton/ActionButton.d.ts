export default ActionButton;
/**
 * Use this component to show an action button.
 */
declare const ActionButton: import('../../../../../node_modules/react').ForwardRefExoticComponent<import('../../../../../node_modules/react').RefAttributes<any>>;
