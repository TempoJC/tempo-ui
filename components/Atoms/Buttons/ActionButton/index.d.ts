import { MouseEventHandler, TouchEventHandler, KeyboardEventHandler, Ref, ReactElement, ReactNode } from '../../../../../node_modules/react';
import { CommonProps } from '../../../../types';
export type ActionButtonProps = CommonProps & {
  autoFocus?: boolean;
  badge?: ReactNode;
  disabled?: boolean;
  form?: string;
  fullWidth?: boolean;
  hideLabel?: boolean;
  iconEnd?: ReactElement;
  iconStart?: ReactElement;
  kind?: "standard" | "danger";
  label?: ReactNode;
  loading?: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onKeyDown?: KeyboardEventHandler<HTMLButtonElement>;
  onKeyUp?: KeyboardEventHandler<HTMLButtonElement>;
  onMouseDown?: MouseEventHandler<HTMLButtonElement>;
  onMouseEnter?: MouseEventHandler<HTMLButtonElement>;
  onMouseLeave?: MouseEventHandler<HTMLButtonElement>;
  onMouseUp?: MouseEventHandler<HTMLButtonElement>;
  onTouchEnd?: TouchEventHandler<HTMLButtonElement>;
  onTouchStart?: TouchEventHandler<HTMLButtonElement>;
  ref?: Ref<HTMLButtonElement>;
  tabIndex?: number;
  title?: string;
  type?: "button" | "submit" | "reset";
};

declare function ActionButton(props: ActionButtonProps): JSX.Element;

export type { ActionButtonProps };
export default ActionButton;
