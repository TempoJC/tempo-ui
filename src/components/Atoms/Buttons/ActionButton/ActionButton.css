.tempo-action-button {
  border: 0;
  outline: 0 none;
  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;
  padding: 15px 0;
  white-space: nowrap;
  max-width: 100%;
  font: var(--typography-label-l-regular-font);
  letter-spacing: var(--typography-label-l-regular-letter-spacing);
  text-transform: var(--typography-label-l-regular-text-transform);
}

.tempo-action-button--badge {
  position: absolute;
  top: calc(var(--size2) - 2px);
  left: calc(100% - var(--size2));
}

.tempo-action-button--icon {
  padding: calc(var(--size3) + 2px) 0;
}

.tempo-action-button--dropdown {
  padding: calc(var(--size3) + 2px) 0;
}

@media (--viewport-l-xl) {
  .tempo-action-button {
    padding: 11px 0;
  }

  .tempo-action-button--icon {
    padding: calc(var(--size3) - 2px) 0;
  }

  .tempo-action-button--dropdown {
    padding: calc(var(--size2) + 2px) 0;
  }

  .tempo-action-button--badge {
    top: calc(var(--size1) - 2px);
  }
}

.tempo-action-button--standard {
  color: var(--color-token-light-content-1);
  background: transparent;
}
.tempo-action-button--color-dark.tempo-action-button--standard {
  color: var(--color-token-dark-content-1);
}

.tempo-action-button--standard:hover {
  cursor: pointer;
  color: var(--color-token-light-content-hover);
}
.tempo-action-button--color-dark.tempo-action-button--standard:hover {
  color: var(--color-token-dark-content-hover);
}

.tempo-action-button--standard:active {
  color: var(--color-token-light-content-active);
}
.tempo-action-button--color-dark.tempo-action-button--standard:active {
  color: var(--color-token-dark-content-active);
}

.tempo-action-button--standard:disabled {
  cursor: default;
  color: var(--color-token-light-content-disabled);
}
.tempo-action-button--color-dark.tempo-action-button--standard:disabled {
  color: var(--color-token-dark-content-disabled);
}

.tempo-action-button--danger {
  color: var(--color-token-light-semantic-danger);
  background: transparent;
}
.tempo-action-button--color-dark.tempo-action-button--danger {
  color: var(--color-token-dark-semantic-danger);
}

.tempo-action-button--danger:hover {
  cursor: pointer;
  color: var(--color-token-light-semantic-danger-hover);
}
.tempo-action-button--color-dark.tempo-action-button--danger:hover {
  color: var(--color-token-dark-semantic-danger-hover);
}

.tempo-action-button--danger:active {
  color: var(--color-token-light-semantic-danger-active);
}
.tempo-action-button--color-dark.tempo-action-button--danger:active {
  color: var(--color-token-dark-semantic-danger-active);
}

.tempo-action-button--danger:disabled {
  cursor: default;
  color: var(--color-token-light-content-disabled);
}
.tempo-action-button--color-dark.tempo-action-button--danger:disabled {
  color: var(--color-token-dark-content-disabled);
}

.tempo-action-button:focus {
  outline: 1px solid var(--color-token-light-semantic-highlight-2);
}

.tempo-action-button:focus:not(:focus-visible) {
  outline: unset;
}

.tempo-action-button--color-dark:focus {
  outline: 1px solid var(--color-token-dark-semantic-highlight-2);
}

.tempo-action-button--color-dark:focus:not(:focus-visible) {
  outline: unset;
}

.tempo-action-button--just-icon:focus {
  outline: 0 none;
}

.tempo-action-button--just-icon .tempo-action-button__loader {
  /* For the loader to have the same size as an icon (20px) */
  padding: 3px;
}

.tempo-action-button--just-icon:focus .tempo-action-button__icon-wrapper,
.tempo-action-button--just-icon:focus .tempo-action-button__loader {
  outline: 1px solid var(--color-token-light-semantic-highlight-2);
}

.tempo-action-button--just-icon:focus:not(:focus-visible) .tempo-action-button__icon-wrapper,
.tempo-action-button--just-icon:focus:not(:focus-visible) .tempo-action-button__loader {
  outline: unset;
}

.tempo-action-button--color-dark.tempo-action-button--just-icon:focus .tempo-action-button__icon-wrapper,
.tempo-action-button--color-dark.tempo-action-button--just-icon:focus .tempo-action-button__loader {
  outline: 1px solid var(--color-token-dark-semantic-highlight-2);
}

.tempo-action-button--color-dark.tempo-action-button--just-icon:focus:not(:focus-visible)
  .tempo-action-button__icon-wrapper,
.tempo-action-button--color-dark.tempo-action-button--just-icon:focus:not(:focus-visible) .tempo-action-button__loader {
  outline: unset;
}

.tempo-action-button__content > * + * {
  margin-left: var(--size1);
}

.tempo-action-button--full-width {
  width: 100%;
  justify-content: center;
}

.tempo-action-button__label {
  overflow: hidden;
  text-overflow: ellipsis;
}

/* Specificity problems... */
.tempo-action-button .tempo-action-button__loader {
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

.tempo-action-button__content {
  display: inline-flex;
  align-items: center;
  max-width: 100%;
}

.tempo-action-button__content--loading {
  visibility: hidden;
}
