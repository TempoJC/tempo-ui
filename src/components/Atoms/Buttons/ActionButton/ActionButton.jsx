import { forwardRef } from "react";
import cn from "classnames";
import { Loader } from "@/components/Feedback/Loader";
import { Icon } from "@/components/Atoms/Icons/Icon";
import "./ActionButton.css";

/**
 * Use this component to show an action button.
 */
const ActionButton = forwardRef(
  (
    {
      as: As = "button",
      className,
      onClick,
      onMouseDown,
      onMouseUp,
      onMouseEnter,
      onMouseLeave,
      onTouchStart,
      onTouchEnd,
      onKeyDown,
      onKeyUp,
      disabled = false,
      fullWidth = false,
      tabIndex,
      autoFocus = false,
      label,
      hideLabel = false,
      iconStart,
      iconEnd,
      testId,
      title,
      type = "button",
      kind = "standard",
      loading = false,
      dropdown = false,
      dropdownVisible = false,
      color = "light",
      badge = null,
      form,
    },
    ref,
  ) => {
    const hasIcon = iconStart || iconEnd;

    return (
      <As
        ref={ref}
        className={cn(className, "tempo-action-button", {
          "tempo-action-button--icon": hasIcon,
          "tempo-action-button--just-icon": hasIcon && !label,
          "tempo-action-button--color-dark": color === "dark",
          "tempo-action-button--full-width": fullWidth,
          "tempo-action-button--standard": kind === "standard",
          "tempo-action-button--danger": kind === "danger",
          "tempo-action-button--dropdown": dropdown,
          "tempo-action-button--loading": loading,
        })}
        onClick={onClick}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onTouchStart={onTouchStart}
        onTouchEnd={onTouchEnd}
        onKeyDown={onKeyDown}
        onKeyUp={onKeyUp}
        disabled={disabled || loading}
        tabIndex={tabIndex}
        autoFocus={autoFocus}
        title={title}
        data-testid={testId}
        type={type}
        form={form}
      >
        {loading && <Loader size="small" color={color} className="tempo-action-button__loader" />}

        <div className={cn("tempo-action-button__content", { "tempo-action-button__content--loading": loading })}>
          {badge && <div className="tempo-action-button--badge">{badge}</div>}
          {iconStart && <span className="tempo-action-button__icon-wrapper">{iconStart}</span>}
          {label && !hideLabel && <span className={"tempo-action-button__label"}>{label}</span>}
          {iconEnd && <span className="tempo-action-button__icon-wrapper">{iconEnd}</span>}
          {dropdown && dropdownVisible && <Icon group="arrows" name="chevron-up-outlined" />}
          {dropdown && !dropdownVisible && <Icon group="arrows" name="chevron-down-outlined" />}
        </div>
      </As>
    );
  },
);

export default ActionButton;
