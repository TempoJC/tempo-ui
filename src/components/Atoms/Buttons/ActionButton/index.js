// packages/components-action/action-button
export { default } from "./ActionButton.jsx";
