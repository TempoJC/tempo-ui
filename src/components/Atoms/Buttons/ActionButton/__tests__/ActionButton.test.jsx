import { userEvent } from "@testing-library/user-event";
import { render, screen, waitFor } from "@testing-library/react";
import ActionButton from "../ActionButton";

describe("@tempo-ui/Atoms/Buttons/ActionButton", () => {
  describe("with icons", () => {
    it("renders the icon before the label when iconStart & label are present", () => {
      const { asFragment } = render(
        <ActionButton className="customClass" label="icon before" iconStart={<div>icon</div>} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders the icon after the label when iconEnd & label are present", () => {
      const { asFragment } = render(
        <ActionButton className="customClass" label="icon after" iconEnd={<div>icon</div>} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("enabled", () => {
    it("renders", async () => {
      const clickHandler = jest.fn();
      const { container } = render(
        <ActionButton
          onClick={clickHandler}
          label="enabled"
          startIcon={<div>some icon</div>}
          testId="action-button-icon"
        />,
      );

      expect(container).toMatchSnapshot();

      const icon = container.querySelector("[data-testid='action-button-icon']");
      await userEvent.click(icon);
      expect(clickHandler).toHaveBeenCalled();
    });
  });

  describe("disabled", () => {
    it("does not call the event handler", async () => {
      const clickHandler = jest.fn();
      const { container } = render(<ActionButton onClick={clickHandler} disabled testId={"action-button-icon"} />);

      const icon = container.querySelector("[data-testid='action-button-icon']");
      await userEvent.click(icon);
      expect(clickHandler).not.toHaveBeenCalled();
    });

    it("renders a loading spinner", async () => {
      const clickHandler = jest.fn();
      render(<ActionButton onClick={clickHandler} loading />);

      await waitFor(() => {
        expect(screen.getByTestId("mock-loader")).toBeInTheDocument();
        expect(screen.getByTestId("mock-loader")).toHaveClass("tempo-action-button__loader");
      });
      await userEvent.click(screen.getByRole("button"));
      expect(clickHandler).not.toHaveBeenCalled();
    });
  });

  describe("renders the dropdown button UI", () => {
    it("without dropdownVisible", () => {
      const { container } = render(<ActionButton dropdown label="Dropdown action button" />);

      expect(container.querySelector("[data-icon-name=chevron-down-outlined]")).toBeInTheDocument();
    });

    it("with dropdownVisible", () => {
      const { container } = render(<ActionButton dropdown dropdownVisible label="Dropdown action button" />);

      expect(container.querySelector("[data-icon-name=chevron-up-outlined]")).toBeInTheDocument();
    });
  });

  it("adds the additional className", () => {
    render(<ActionButton className="customClass" label="custom" />);
    expect(screen.getByRole("button")).toHaveClass("customClass");
  });

  describe("form", () => {
    it("sets the form attribute if provided", async () => {
      render(<ActionButton form="my-form" label="custom" />);

      expect(screen.getByRole("button")).toHaveAttribute("form", "my-form");
    });
  });

  describe("testId prop", () => {
    it("renders the ActionButton with the specified data-testid attribute", () => {
      const { getByTestId } = render(<ActionButton label="button-title" testId="action-button" />);
      const buttonElement = getByTestId("action-button");
      expect(buttonElement).toBeInTheDocument();
      expect(buttonElement).toHaveAttribute("data-testid", "action-button");
    });
  });
});
