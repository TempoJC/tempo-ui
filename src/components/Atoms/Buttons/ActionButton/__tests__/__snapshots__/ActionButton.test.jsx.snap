// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/Atoms/Buttons/ActionButton enabled renders 1`] = `
<div>
  <button
    class="tempo-action-button tempo-action-button--standard"
    data-testid="action-button-icon"
    type="button"
  >
    <div
      class="tempo-action-button__content"
    >
      <span
        class="tempo-action-button__label"
      >
        enabled
      </span>
    </div>
  </button>
</div>
`;

exports[`@tempo-ui/Atoms/Buttons/ActionButton with icons renders the icon after the label when iconEnd & label are present 1`] = `
<DocumentFragment>
  <button
    class="customClass tempo-action-button tempo-action-button--icon tempo-action-button--standard"
    type="button"
  >
    <div
      class="tempo-action-button__content"
    >
      <span
        class="tempo-action-button__label"
      >
        icon after
      </span>
      <span
        class="tempo-action-button__icon-wrapper"
      >
        <div>
          icon
        </div>
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`@tempo-ui/Atoms/Buttons/ActionButton with icons renders the icon before the label when iconStart & label are present 1`] = `
<DocumentFragment>
  <button
    class="customClass tempo-action-button tempo-action-button--icon tempo-action-button--standard"
    type="button"
  >
    <div
      class="tempo-action-button__content"
    >
      <span
        class="tempo-action-button__icon-wrapper"
      >
        <div>
          icon
        </div>
      </span>
      <span
        class="tempo-action-button__label"
      >
        icon before
      </span>
    </div>
  </button>
</DocumentFragment>
`;
